KAFKA-12554: Refactor Log layer

kowshik opened the pull request:
**TL;DR:**

This PR implements the details of the Log layer refactor, as outlined in this document: https://docs.google.com/document/d/1dQJL4MCwqQJSPmZkVmVzshFZKuFy_bCPtubav4wBfHQ/edit. Few details maybe different from the doc, but it is more or less the same.

**STRATEGY:**

In this PR, I've extracted a new class called `LocalLog` out of `Log`.  Currently `LocalLog` is purely an implementation detail thats not exposed outside `Log` class (except for tests). The object encapsulation is that each `Log` instance  wraps around a `LocalLog` instance.

This new `LocalLog` class attempts to encompass **most** of the responsibilities of local log surrounding the `segments` map, which otherwise were present in `Log` previously. Note that **not** all local log responsibilities have been moved over to this new class (yet). The criteria I used was to preserve (for now) in existing `Log` class, any logic that is mingled  in a complex manner with the `logStartOffset` or the `LeaderEpochCache` or the `ProducerStateManager`.

**WINS:**

The main win is that the new `LocalLog` class is now agnostic of the `logStartOffset`, which continues to be managed mainly by `Log` class. Below is the local log functionality that has successfully moved over from `Log` to `LocalLog`:

1. Access of `LogSegments` instance containing the local `LogSegment` objects.
2. Read path logic to read records from the log.
2. Segment file deletion logic.
3. Segment truncation logic.
4. Segment roll logic.
5. Segment split logic.
6. Segment replacement logic.

Below is the main local log functionality that continues to remain in `Log`:

1. Segment append logic. The reason is that the below logic is mingled with one or more of the following: `logStartOffset` or `LeaderEpochCache` or `ProducerStateManager`. This makes it hard to separate just the local logic out of it.
2. Last stable offset and logic surrounding it.
3. High watermark and logic surrounding it.
4. Logic to `fetchOffsetByTimestamp` and logic to `legacyFetchOffsetsBefore`.
5. Some of the retention logic thats related with the global view of the log.
6. All other logic related with handling  `LeaderEpochCache` and `ProducerStateManager`.

**PAINPOINTS:**

1. Log locking semantics had to be changed in handful of areas, with lock taken at a coarse level.
2. Few API implementations needed re-ordering of logic in `Log` class to make migration feasible.
3. Certain APIs added to `LocalLog` are crude in nature or signature, examples: `def checkIfMemoryMappedBufferClosed`, `def markFlushed`, `def updateRecoveryPoint`, `def replaceSegments` etc.
4. Certain important APIs (such as `def append` logic) were hard to migrate because logic was mingled with Leader epoch cache, Producer state manager and log start offset.

**TESTS:**
 * New unit suite: `LocalLogTest.scala` has been provided containing tests specific to `LocalLog` class.
All other existing tests are expected to pass.
 * 6/10/2021: System test run on top of 28bf22af168ca0db76796b5d3cd67a38ed8ed1c2: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/4550/
 *  6/12/2021:
    * System test runs [4560](https://jenkins.confluent.io/job/system-test-kafka-branch-builder/4560/) and [4562](https://jenkins.confluent.io/job/system-test-kafka-branch-builder/4562/) on top of 008b701386ce5a4d892d6ac5b90798b981c4fba0.
    * System test runs [4561](https://jenkins.confluent.io/job/system-test-kafka-branch-builder/4561/) and [4563](https://jenkins.confluent.io/job/system-test-kafka-branch-builder/4563/) on top of `trunk/6de37e536ac76ef13530d49dc7320110332cd1ee`.
    * `kafkatest.tests.client.consumer_test` rerun:
        * [4564](https://jenkins.confluent.io/job/system-test-kafka-branch-builder/4564/) against  `trunk/6de37e536ac76ef13530d49dc7320110332cd1ee`.
        * [4566](https://jenkins.confluent.io/job/system-test-kafka-branch-builder/4566/) on top of 008b701386ce5a4d892d6ac5b90798b981c4fba0.

ijuma commented:
Can we do one PR for renaming `Log` to `LocalLog` and then a separate ones for the rest? It seems like git rename detection failed here and it will make diffs harder to review.

kowshik commented:
@ijuma In this PR the intention was not to rename `Log` to `LocalLog`, but rather extract `LocalLog` class out of `Log`. My current plan is to do the alternative of what you had suggested i.e. extract `LocalLog` from `Log` in this PR and then rename `Log` to `GlobalLog` in a subsequent PR. The reason is to focus on the more important/major piece in the first PR (which is the Log layer separation), then the renaming the abstractions in a future PR becomes a relatively minor activity. Either way we choose, we will eventually introduce a new abstraction viz. `LocalLog` or `GlobalLog`.

Thoughts?

ijuma commented:
OK, thanks for the explanation. Btw, why do we call one of the logs `GlobalLog`? In what sense is it `Global`?

kowshik commented:
@ijuma The purpose of `GlobalLog` class is to serve as a higher layer, stitching together the unified view of both the local and remote portion of the log. Importantly, this class is aware of the global log start offset. This class is external facing, and acts as the outer shell, meaning that the public API of this class will be used by other components such as LogManager, LogCleaner etc. and other components outside the kafka.log package. It could just be called as `Log` too, but I thought by calling it `GlobalLog` the intention/differentiation is clear.

The above and few more things are explained in the doc attached in the description. I'd suggest having a look at the doc.

ijuma commented:
@kowshik I can't comment on the doc, that's why I commented here. :) I didn't see any reason there for calling it `Global` btw. `Global` tends to imply something more than what this is doing IMO.

kowshik commented:
@ijuma I've opened up the doc for comments. I've also updated it to use the name `UnifiedLog` instead of `GlobalLog`. Hopefully the intent is better communicated now in the naming.

junrao commented on core/src/main/scala/kafka/log/LocalLog.scala:
```diff
@@ -0,0 +1,1561 @@
+
+/**
+ * An append-only log for storing messages locally.
+ * The log is a sequence of LogSegments, each with a base offset denoting the first message in the segment.
```
"each with a base offset denoting the first message in the segment" : This is probably not true for a compacted segment. Perhaps we could just say base offset.

kowshik replied:
Done.

junrao commented on core/src/main/scala/kafka/log/LocalLog.scala:
```diff
@@ -0,0 +1,1561 @@
+  }
+
+  /**
+   * Add the given segment to the segments in this log. If this segment replaces an existing segment, delete it.
```
The comment is inaccurate since the method no longer deletes segments.

kowshik replied:
Done. Removed.

junrao commented on core/src/main/scala/kafka/log/LocalLog.scala:
```diff
@@ -0,0 +1,1561 @@
+  /**
+   * Clears all segments
+   */
+  private[log] def clearSegments(): Unit = segments.clear()
```
This seems unused in Log.

kowshik replied:
Done. Removed.

junrao commented on core/src/main/scala/kafka/log/LocalLog.scala:
```diff
@@ -0,0 +1,1561 @@
+  /**
+   * Closes all segments
+   */
+  private[log] def closeSegments(): Unit = {
```
This seems only used in `close()`? If so, could we just fold it into `close()`?

kowshik replied:
Done. I've moved it into`close()` now. Good point.

junrao commented on core/src/main/scala/kafka/log/LocalLog.scala:
```diff
@@ -0,0 +1,1561 @@
+   *
+   * @throws LogSegmentOffsetOverflowException if we encountered a legacy segment with offset overflow
+   */
+  private[log] def recoverLog(logStartOffset: Long,
```
This is only used inside localLog and doesn't need package level access.

kowshik replied:
Done. Good point. I've made it private now.

junrao commented on core/src/main/scala/kafka/log/LocalLog.scala:
```diff
@@ -0,0 +1,1561 @@
+  }
+
+  /**
+   * Recover the log segments and return the next offset after recovery.
```
This method returns more than the next offset.

kowshik replied:
Done. Updated the doc now.

junrao commented on core/src/main/scala/kafka/log/LocalLog.scala:
```diff
@@ -0,0 +1,1561 @@
+
+    // if we have the clean shutdown marker, skip recovery
+    if (!hadCleanShutdown) {
+      val unflushed = logSegments(this.recoveryPoint, Long.MaxValue).iterator
```
No need for `this`. Ditto in a few other places.

kowshik replied:
Done. Removed now.

junrao commented on core/src/main/scala/kafka/log/LocalLog.scala:
```diff
@@ -0,0 +1,1561 @@
+   *
+   * @throws LogSegmentOffsetOverflowException if the log directory contains a segment with messages that overflow the index offset
+   */
+  private[log] def loadSegmentFiles(logStartOffset: Long, maxProducerIdExpirationMs: Int): Unit = {
```
This is only used inside localLog and doesn't need package level access.

kowshik replied:
Done. Good point. I've fixed it now.

junrao commented on core/src/main/scala/kafka/log/LocalLog.scala:
```diff
@@ -0,0 +1,1561 @@
+   *                                           this situation. This is expected to be an extremely rare scenario in practice,
+   *                                           and manual intervention might be required to get out of it.
+   */
+  private[log] def completeSwapOperations(swapFiles: Set[File],
```
This is only used inside localLog and doesn't need package level access.

kowshik replied:
Done. Good point. I've fixed it now.

junrao commented on core/src/main/scala/kafka/log/LocalLog.scala:
```diff
@@ -0,0 +1,1561 @@
+   *
+   * @return Set of .swap files that are valid to be swapped in as segment files
+   */
+  private[log] def removeTempFilesAndCollectSwapFiles(): Set[File] = {
```
This is only used inside localLog and doesn't need package level access.

kowshik replied:
Done. Good point. I've fixed it now.

junrao commented on core/src/main/scala/kafka/log/LocalLog.scala:
```diff
@@ -0,0 +1,1561 @@
+    validSwapFiles
+  }
+
+  private[log] def retryOnOffsetOverflow[T](fn: => T): T = {
```
This is only used inside localLog and doesn't need package level access. Ditto in a few other places.

kowshik replied:
Done. Good point. I've fixed it now.

junrao commented on core/src/main/scala/kafka/log/LocalLog.scala:
```diff
@@ -0,0 +1,1561 @@
+  /**
+   * Perform physical deletion for the given segments. Allows the segments to be deleted asynchronously or synchronously.
+   *
+   * This method assumes that the segment exists and the method is not thread-safe.
```
The whole class is not thread-safe.

kowshik replied:
Done. Fixed the comment now.

junrao commented on core/src/main/scala/kafka/log/LocalLog.scala:
```diff
@@ -0,0 +1,1561 @@
+   *
+   * This method assumes that the segment exists and the method is not thread-safe.
+   *
+   * This method does not need to convert IOException (thrown from changeFileSuffixes) to KafkaStorageException because
```
This is an existing problem, but the comment is inaccurate since it does convert IOException to KafkaStorageException.

kowshik replied:
The method only converts `IOException` to `KafkaStorageException` inside the internal `deleteSegments()` helper function, but it doesn't do the same for `changeFileSuffixes` which is what the comment was referring to. I've improved the `@throws` doc to refer to both exceptions now. Please let me know if it needs further improvement.

junrao commented on core/src/main/scala/kafka/log/LocalLog.scala:
```diff
@@ -0,0 +1,1561 @@
+   * Asynchronous deletion allows reads to happen concurrently without synchronization and without the possibility of
+   * physically deleting a file while it is being read.
+   *
+   * This method does not need to convert IOException to KafkaStorageException because it is either called before all logs are loaded
```
This is an existing problem, but the comment is inaccurate since it does convert IOException to KafkaStorageException.

kowshik replied:
I've updated the doc now. The reason is the same the one I mentioned in the [above response](https://github.com/apache/kafka/pull/10280/files#r602052452).

junrao commented on core/src/main/scala/kafka/log/LocalLog.scala:
```diff
@@ -0,0 +1,1561 @@
+  }
+
+  /**
+   * This function does not acquire Log.lock. The caller has to make sure log segments don't get deleted during
```
Log.lock is in a different class now.

kowshik replied:
Done. I've fixed it now.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -213,32 +208,38 @@ case object SnapshotGenerated extends LogStartOffsetIncrementReason {
- * The log is a sequence of LogSegments, each with a base offset denoting the first message in the segment.
+ * The log consists of tiered and local segments with the tiered portion of the log being optional. There could be an
+ * overlap between the tiered and local segments. The active segment is always guaranteed to be local. If tiered segments
+ * are present, they always appear at the head of the log, followed by an optional region of overlap, followed by the local
```
Some people think the end of the log is the head. So, maybe use beginning/end of log instead.

kowshik replied:
Done. Good point. I've fixed the doc to refer to `beginning of log`.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -213,32 +208,38 @@ case object SnapshotGenerated extends LogStartOffsetIncrementReason {
- * New log segments are created according to a configurable policy that controls the size in bytes or time interval
- * for a given segment.
+ * NOTE: this class handles state and behavior specific to tiered segments as well as any behavior combining both tiered
+ * and local segments. The state and behavior specific to local segments is handled by the encapsulated LocalLog instance.
```
is handled => are handled

kowshik replied:
Done.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -836,178 +581,15 @@ class Log(@volatile private var _dir: File,
 
   private def loadProducerState(lastOffset: Long, reloadFromCleanShutdown: Boolean): Unit = lock synchronized {
-    rebuildProducerState(lastOffset, reloadFromCleanShutdown, producerStateManager)
+    lock synchronized {
```
Hmm, we don't need to synchronize on the lock twice, right?

kowshik replied:
Done. Good catch.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -1794,20 +1269,21 @@ class Log(@volatile private var _dir: File,
-      isolation = FetchLogEnd,
-      minOneMessage = false)
-    fetchDataInfo.fetchOffsetMetadata
+  def convertToOffsetMetadata(offset: Long): Option[LogOffsetMetadata] = {
```
Is this method being used?

kowshik replied:
Done. Removed now. Good catch.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -1816,8 +1292,12 @@ class Log(@volatile private var _dir: File,
   private def deleteOldSegments(predicate: (LogSegment, Option[LogSegment]) => Boolean,
                                 reason: SegmentDeletionReason): Int = {
+    def shouldDelete(segment: LogSegment, nextSegmentOpt: Option[LogSegment], logEndOffset: Long): Boolean = {
+      highWatermark >= nextSegmentOpt.map(_.baseOffset).getOrElse(logEndOffset) &&
```
Hmm, why do we need to wrap `predicate` with an additional condition?

kowshik replied:
This is to accomodate for the hwm check that was previously happening in `Log#deletableSegments` in [this line](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/log/Log.scala#L1872). The `deletableSegments` method has now moved to `LocalLog`, but we can't do the hwm check inside `LocalLog` since hwm is still owned by `Log`. So we piggyback on the predicate here to additionally attach the hwm check.

junrao reviewed (commented):
@kowshik : Thanks for the PR. A few comments below.

kowshik commented:
@junrao Thanks a lot for the review! I've addressed your comments in 63be325b8b3ff5514e22137bb02dfc70e10f7a72.

kowshik commented:
@junrao Just a heads up on the following. I'm working on the changes for KAFKA-12552 (https://github.com/apache/kafka/pull/10401 to extract segments map) and KAFKA-12553 (refactor recovery logic) in separate PRs. It seems better if we merge those into trunk ahead of the current PR.

kowshik committed:
KAFKA-12554: Refactor Log layer

kowshik commented:
@junrao This PR is ready for another round of review. I've rebased the PR onto latest AK trunk, iterated on the implementation bit more and added new unit tests for `LocalLog` class under `LocalLogTest.scala`.

kowshik commented:
@junrao This PR is ready for another round of review. I've rebased the PR onto latest AK trunk, iterated on the implementation bit more and added new unit tests for `LocalLog` class under `LocalLogTest.scala`.

cc @dhruvilshah3

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -1500,50 +1325,67 @@ class Log(@volatile private var _dir: File,
+      producerStateManager.takeSnapshot()
+    },
+    postRollAction = (newSegment: LogSegment, deletedSegment: Option[LogSegment]) => {
+      deletedSegment.foreach(segment => deleteProducerSnapshotAsync(Seq(segment)))
```
This seems to have exposed an existing bug. During roll, deletedSegment will be non-empty if there is an existing segment of 0 size with the newOffsetToRoll. However, since we take a producer snapshot on newOffsetToRoll before calling postRollAction, we will be deleting the same snapshot we just created.

In this case, I think we don't need to delete producerSnapshot for deletedSegment.

kowshik replied:
This is a great catch. I agree with you. While I can address it in this PR, should we create a separate JIRA for it?

junrao replied:
We could fix this in a separate jira too.

kowshik replied:
I've created a JIRA to track this: https://issues.apache.org/jira/browse/KAFKA-12876.

kowshik replied:
Done.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -1500,50 +1325,67 @@ class Log(@volatile private var _dir: File,
-  def logEndOffset: Long = nextOffsetMetadata.messageOffset
+  def logEndOffsetMetadata: LogOffsetMetadata = localLog.logEndOffsetMetadata
+
+  private val rollAction = RollAction(
```
I feel RollAction actually makes the code harder to understand than before. So, it would be useful to think through if we could avoid it. In particular, it seems that anything in postRollAction could just be done in the caller if we return enough context. We are taking a producer snapshot in preRollAction. However, since we are not adding new data here. It seems that we could take producer snapshot in Log.roll() after calling localLog.roll() while holding the Log.lock.

kowshik replied:
Great point. I agree with you that `RollAction` is avoidable and it complicates the code a bit. The main part I wasn't sure about was why previously in the code we chose to take the producer state snapshot before the segment was rolled, instead of doing it after. Am I right that taking the snapshot prior to the roll is not a strict requirement for crash safety? (I'm not able to think of a case myself, but wanted to double check)

junrao replied:
Right, I don't think there is a particular reason that we have to take a snapshot before adding a new empty segment.

kowshik replied:
Done.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -1572,144 +1414,69 @@ class Log(@volatile private var _dir: File,
-        // File handlers will be closed if this log is deleted
-        isMemoryMappedBufferClosed = true
+        val deletedSegments = localLog.delete()
+        deleteProducerSnapshotAsync(deletedSegments)
```
Hmm, the ordering is a bit weird. We delete the directory in localLog.delete(). However, the producer snapshot is in the directory and is deleted later.

kowshik replied:
That's a good point. The best solution I could think of is to split the functionality up into few different APIs in `LocalLog`. For example, this is how the right implementation could look like in `Log.scala`:

```
// In `Log.scala`:
private[log] def delete(): Unit = {
  maybeHandleIOException(s"Error while deleting log for $topicPartition in dir ${dir.getParent}") {
    lock synchronized {
      producerExpireCheck.cancel(true)
      leaderEpochCache.foreach(_.clear())
      val deletedSegments = localLog.deleteAllSegments()
      deleteProducerSnapshotAsync(deletedSegments)
      localLog.deleteDir()
    }
  }
}
  
```

kowshik replied:
Done.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -1806,37 +1566,37 @@ class Log(@volatile private var _dir: File,
-    rebuildProducerState(endOffset, producerStateManager)
+    lock synchronized {
+      rebuildProducerState(endOffset, producerStateManager)
+    }
```
This change has a couple of issues. 
(1) updateHighWatermark() now only updates the offset, but not the corresponding offset metadata. The offset metadata is needed in serving fetch requests. Recomputing that requires index lookup and log scan, and can be extensive. So, we need to preserve the offset metadata during truncate() and truncateFully().
(2) I think updateHighWatermark() needs to be called within the lock. updateHighWatermark() reads local log's logEndOffset. So, we don't want the logEndOffset to change while updateHighWatermark() is called.

kowshik replied:
Sounds good. I'll fix this.

kowshik replied:
Done.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -1852,65 +1612,24 @@ class Log(@volatile private var _dir: File,
-      config, scheduler, logDirFailureChannel, producerStateManager, this.logIdent)
-  }
-
   private[log] def replaceSegments(newSegments: Seq[LogSegment], oldSegments: Seq[LogSegment], isRecoveredSwapFile: Boolean = false): Unit = {
```
It seems that we could remove isRecoveredSwapFile since it's always false from the caller.

kowshik replied:
Great catch, I'll fix this.

kowshik replied:
Done.

junrao commented on core/src/main/scala/kafka/log/LogLoader.scala:
```diff
@@ -246,17 +262,17 @@ object LogLoader extends Logging {
             params.logDirFailureChannel,
-            params.producerStateManager,
             params.logIdentifier)
+          deleteProducerSnapshotsAsync(result.deletedSegments, params)
```
This is unnecessary since during splitting, the old segment is replaced with a new segment with the same base offset. So, result.deletedSegments is always empty.

kowshik replied:
Sounds good. Great catch. It appears straightforward to just skip deleting the snapshot here, I can leave a comment explaining why.

kowshik replied:
@junrao I thought about this again. Correct me if I'm wrong, but it appears we may be altering existing behavior if we go down this route. Should we do it in a separate PR to isolate the change?

junrao replied:
Yes, that's fine.

kowshik replied:
I have created a jira to track this improvement. https://issues.apache.org/jira/browse/KAFKA-12923

kowshik replied:
I've created a jira to track this. https://issues.apache.org/jira/browse/KAFKA-12923

junrao commented on core/src/test/scala/unit/kafka/log/LocalLogTest.scala:
```diff
@@ -0,0 +1,734 @@
+
+  @AfterEach
+  def tearDown(): Unit = {
+    if (!log.isMemoryMappedBufferClosed) {
```
Should we use log.checkIfMemoryMappedBufferClosed()?

kowshik replied:
Is the concern here that we are accessing the private attribute?
In certain tests, `log.checkIfMemoryMappedBufferClosed()` will raise an exception if the log has already been closed. If the private attribute access is a concern, I can change this to just call `log.checkIfMemoryMappedBufferClosed()` but ignore the `KafkaStorageException` thats raised.

kowshik replied:
Done.

kowshik replied:
Done.

junrao commented on core/src/test/scala/unit/kafka/log/LocalLogTest.scala:
```diff
@@ -0,0 +1,734 @@
+    val oldConfig = log.config
+    assertEquals(oldConfig, log.config)
+
+    val newConfig = createLogConfig()
```
It seems newConfig is always the same as oldConfig?

kowshik replied:
That's a good point. While it is still a different reference, we can still create a new log config with an override, something like: `createLogConfig(segmentBytes=oldConfig.segmentSize * 2)`

kowshik replied:
Done.

junrao commented on core/src/test/scala/unit/kafka/log/LocalLogTest.scala:
```diff
@@ -0,0 +1,734 @@
+        assertEquals(log.logEndOffset, logEndOffset)
+        offset += 1
+        true
+      })
```
Should we assert sth after the log.deletableSegments() call?

kowshik replied:
Good question. We could assert that the returned value matches our expectations. Also we could assert that `log.deletableSegments()` does not make any mutations to the `LocalLog` object.

kowshik replied:
Done.

junrao reviewed (commented):
@kowshik : Thanks for the updated PR. A few more comments below.

kowshik committed:
Address comments

kowshik commented:
@junrao Thanks for the review! I've addressed the comments in e201295e03e0ea8a7102983888d1a7afc66d384a, and have also rebased this PR onto most recent commit in `trunk`. It is ready for review again.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -1578,144 +1409,85 @@ class Log(@volatile private var _dir: File,
-      roll(Some(rollOffset))
+      val newSegment = localLog.roll(Some(rollOffset))
+      afterRoll(newSegment)
+      newSegment
```
Could we replace the above 3 lines with `roll(Some(rollOffset))`?

kowshik replied:
Sure.

kowshik replied:
Done in 8f14879.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -1578,144 +1409,85 @@ class Log(@volatile private var _dir: File,
 
   /**
-   * Roll the log over to a new active segment starting with the current logEndOffset.
+   * Roll the local log over to a new active segment starting with the current logEndOffset.
```
This comment is not very accurate since we roll to expectedNextOffset or logEndOffset.

kowshik replied:
Sure, I'll fix it. Good catch.

kowshik replied:
Done in 8f14879.

junrao commented on core/src/main/scala/kafka/log/LocalLog.scala:
```diff
@@ -0,0 +1,1010 @@
+      }
+      Utils.delete(dir)
+      // File handlers will be closed if this log is deleted
+      isMemoryMappedBufferClosed = true
```
It seems that we should set isMemoryMappedBufferClosed in deleteAllSegments()?

kowshik replied:
That's a good point. I'll move it there.

kowshik replied:
Done in 8f14879.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -1812,37 +1577,36 @@ class Log(@volatile private var _dir: File,
     logStartOffset = startOffset
-    nextOffsetMetadata = LogOffsetMetadata(endOffset, activeSegment.baseOffset, activeSegment.size)
-    recoveryPoint = math.min(recoveryPoint, endOffset)
+    localLog.updateLogEndOffset(endOffset)
```
We need to preserve the LogOffsetMetadata for endOffset and use it to call updateHighWatermark.

kowshik replied:
Sounds good. This can be updated to `updateHighWatermark(localLog.logEndOffsetMetadata)`.

kowshik replied:
Done in 8f14879.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -1788,14 +1559,8 @@ class Log(@volatile private var _dir: File,
-          initFileSize = config.initFileSize,
-          preallocate = config.preallocate))
+        val deletedSegments = localLog.truncateFullyAndStartAt(newOffset)
+        deleteProducerSnapshots(deletedSegments, asyncDelete = true)
```
`producerStateManager.truncateFullyAndStartAt()` removes all producer snapshots. So, this is necessary.

kowshik replied:
Sounds good. I'll fix this.

kowshik replied:
Done in 8f14879.

junrao reviewed (commented):
@kowshik : Thanks for the updated PR. A few more comments. Also, could you run all system tests for the PR?

kowshik committed:
Address comments

kowshik commented:
Thanks for the review @junrao! I've addressed the comments in 8ebb39f.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -1812,37 +1570,39 @@ class Log(@volatile private var _dir: File,
+    localLog.updateLogEndOffset(endOffset)
     rebuildProducerState(endOffset, producerStateManager)
-    updateHighWatermark(math.min(highWatermark, endOffset))
+    if (highWatermark < localLog.logEndOffset)
```
If highWatermark is smaller, there is no need to update high watermark.

kowshik replied:
Done in 28bf22af168ca0db76796b5d3cd67a38ed8ed1c2.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -1788,17 +1554,9 @@ class Log(@volatile private var _dir: File,
-          time = time,
-          initFileSize = config.initFileSize,
-          preallocate = config.preallocate))
+        localLog.truncateFullyAndStartAt(newOffset)
```
This is an existing issue. In this case, it seems that we should always update high watermark in completeTruncation() with localLog.logEndOffsetMetadata.

kowshik replied:
Done in 28bf22af168ca0db76796b5d3cd67a38ed8ed1c2.

junrao reviewed (commented):
@kowshik : Thanks for the updated PR. Just a couple of more comments.

kowshik committed:
Address comments from Jun

kowshik commented:
Thanks for the review @junrao! I've addressed your most recent comments in 28bf22af168ca0db76796b5d3cd67a38ed8ed1c2.

junrao reviewed (approved):
@kowshik : Thanks for the updated PR. LGTM. Do you have the system test results?

kowshik commented:
@junrao Thanks for the review! I ran the system tests.
1. [System test run #4560](https://jenkins.confluent.io/job/system-test-kafka-branch-builder/4560/) on top of the latest commit 008b701386ce5a4d892d6ac5b90798b981c4fba0 from this PR. The run finished with 12 test failures.
2. [System test run #4561](https://jenkins.confluent.io/job/system-test-kafka-branch-builder/4561/) against AK trunk on top of commit 6de37e536ac76ef13530d49dc7320110332cd1ee which does not contain changes from this PR. The run finished with 13 test failures.

There were 11 overlapping failures in both (1) and (2). For these, I didn't find anything abnormal in the logs so far, the failure reason seems similar in both.

The only new failure in (1) that's not present in (2) was:

```
Module: kafkatest.tests.client.consumer_test
Class:  OffsetValidationTest
Method: test_broker_failure
Arguments:
{
  "clean_shutdown": true,
  "enable_autocommit": false,
  "metadata_quorum": "REMOTE_KRAFT"
}
```

Logs indicate that the test failed [at this line](https://github.com/apache/kafka/blob/b96fc7892f1e885239d3290cf509e1d1bb41e7db/tests/kafkatest/tests/client/consumer_test.py#L388) because one of the worker nodes running the consumer didn't complete within the timeout of 30s. This doesn't seem indicative of a real failure (yet), so I'm rerunning the system tests again in [test run #4562](https://jenkins.confluent.io/job/system-test-kafka-branch-builder/4561/) to check if the failure is consistent.

satishd reviewed (approved):
Thanks @kowshik for the PR. I had a look at LocalLog and Log classes and their dependent changes, it looks good to me. +1

kowshik commented:
I ran the system tests in `kafkatest.tests.client.consumer_test` again:
 * [4564](https://jenkins.confluent.io/job/system-test-kafka-branch-builder/4564/) against  `trunk/6de37e536ac76ef13530d49dc7320110332cd1ee`.
 * [4566](https://jenkins.confluent.io/job/system-test-kafka-branch-builder/4566/) against 008b701386ce5a4d892d6ac5b90798b981c4fba0 (the latest commit from this PR).

All tests passed.

kowshik commented:
@junrao @dhruvilshah3 I ran a perf test against a Broker build with and without this PR. The test involved the following:

1. Created a test topic with 1 partition and replication factor 1 using the command: `$> ./bin/kafka-topics.sh --bootstrap-server localhost:9092 --create --topic kowshik-test-1 --partitions 1 --replication-factor 1`.
2. Ran `kafka-producer-perf-test.sh` to produce 10M messages each of size 1KB and with max producer throughput 100K to the above topic. Command: `$> ./bin/kafka-producer-perf-test.sh --num-records 10000000 --print-metrics --producer-props bootstrap.servers=localhost:9092 --record-size 1024 --throughput 100000 --topic kowshik-test-1`.
3. In parallel, ran `kafka-consumer-perf-test.sh` to consume the 10M messages that were produced in (2) using the command: `$> bin/kafka-consumer-perf-test.sh --topic kowshik-test-1 --bootstrap-server localhost:9092 --messages 10000000 --print-metrics --show-detailed-stats`

The tests have similar results, meaning that the performance with and without this PR looks similar. Here are the results:


**log.segment.bytes=10MB**

- Without this PR on top of c333bfd41766bf33f41b5d32d8959ebbeff240b4: https://gist.github.com/kowshik/0ea1ae9ac8210f4bba49967727ddb475
- With this PR on top of c333bfd41766bf33f41b5d32d8959ebbeff240b4: https://gist.github.com/kowshik/c5ec0fb92679bd91613f520455446bf5

**log.segment.bytes=100MB**

- Without this PR on top of c333bfd41766bf33f41b5d32d8959ebbeff240b4: https://gist.github.com/kowshik/2aaa7113fd05e10721c60aaf9bf8c654
- With this PR on top of c333bfd41766bf33f41b5d32d8959ebbeff240b4: https://gist.github.com/kowshik/b09882c2b13930be2efc69554c31aded

**log.segment.bytes=1GB**

- Without this PR on top of c333bfd41766bf33f41b5d32d8959ebbeff240b4: https://gist.github.com/kowshik/ea36153c9751180c5dbe383b189d50df
- With this PR on top of c333bfd41766bf33f41b5d32d8959ebbeff240b4: https://gist.github.com/kowshik/133803d10d510df93f1d15858e91035a

kowshik committed:
Minor cleanups + changes

junrao reviewed (approved):
@kowshik : Thanks for the updated PR. LGTM. Are there any more tests that you plan to do?

kowshik committed:
Avoid reference to Log inside LocalLog

kowshik committed:
Minor logging improvements

kowshik commented:
@junrao Thanks for the review. I ran load tests on the changes from this PR, there weren't any new regressions (i.e. latency regressions or errors) that I noticed, except for an issue that I found which looks unrelated to this PR, its described in this jira: https://issues.apache.org/jira/browse/KAFKA-13070.

The load test was run on a 6-broker cluster with 250GB SSD disks:
 * Produce consume on a test topic 2000 partitions (~1000+ replica count per broker).
 * Per topic # of producers = 6.
 * Produce ingress per broker = ~20.5MBps.
 * Per topic # of consumers = 6.
 * \# of consumer groups = 3.
 * Test duration: ~1h.

Mid-way through the test, I rolled the cluster under load to check how the cluster behaved. Overall things looked OK.

satishd commented:
Thanks @junrao for merging into trunk. Can we also push this to [3.0 branch](https://github.com/apache/kafka/tree/3.0) as we discussed earlier?
cc @kowshik

ijuma commented:
What is the reason for including a refactoring in 3.0 after the feature freeze?

kowshik commented:
@ijuma Discussed with @satishd. We are not planning to include this in 3.0.
